[PF-369] Add retry to all database requests

ddietterich opened the pull request:
There are two main parts of this PR. First, I created StairwayInstanceDao to handle interactions with the StairwayInstance table. The FlightDao was getting large and those methods separate fairly cleanly. That required factoring out the common transaction methods into DbUtils class.

Second, I made a generic retry class (DbRetry) and applied it across both DAOs.

This should allow TDR to ride through transient SqlProxy failures and serialization contention.

The Stairway unit tests pass. I also ran the Data Repo connected tests to further sanity check.

ddietterich committed:
Factor stairway instance DAO from flight DAO
Make shared utility methods (DbUtils)
Implement generic retry class (DbRetry)
Apply retry in all cases

pshapiro4broad commented on src/main/java/bio/terra/stairway/FlightDao.java:
```diff
@@ -55,144 +56,43 @@
-  /** Record a new flight */
   void submit(FlightContext flightContext)
       throws DatabaseOperationException, InterruptedException, DuplicateFlightIdSubmittedException {
+    DbRetry dbRetry = new DbRetry("flight.submit");
```
For the pattern `new Foo().method()` (that is, the case where you create an object for the sole purpose of calling a single method on it) I think it's clearer to use a static method on the object. You can still create an object in that static method, but the caller doesn't need to know about this detail.

pshapiro4broad commented on src/main/java/bio/terra/stairway/DbRetry.java:
```diff
@@ -0,0 +1,69 @@
+  private void retryWait() throws InterruptedException {
+    int sleepMS = ThreadLocalRandom.current().nextInt(WAIT_MIN_MS, WAIT_MAX_MS);
+    TimeUnit.MILLISECONDS.sleep(sleepMS);
+    logger.info("retrying for {}", logString);
```
IMO this seems like `debug` not `info`. Unless you believe retrying will be relatively uncommon.

pshapiro4broad commented on src/main/java/bio/terra/stairway/DbFunction.java:
```diff
@@ -0,0 +1,14 @@
+ * @param <R> The return value of the database function. If there is no return value, specify Void.
+ */
+@FunctionalInterface
+public interface DbFunction<R> {
```
Since this is only used in the class `DbRetry` I don't think there's a benefit to declaring in a top level scope. Declaring it local to `DbRetry` would have the same effect and not pollute the global namespace.

pshapiro4broad commented on src/main/java/bio/terra/stairway/FlightDao.java:
```diff
@@ -55,144 +56,43 @@
+    dbRetry.perform(() -> submitInner(flightContext));
+  }
+
+  private Void submitInner(FlightContext flightContext)
```
I'm not sure how well this would work, but you could have two `perform` methods, with one that supports `void` methods. That would avoid the need for `Void` and `return null;`

ddietterich replied:
From what I read, an `@FunctionalInterface` has to have a return value. The recommendation was to return Void, but you cannot `new` a Void, so you return null.

Is there a variant of functional interface that does not return a value?

pshapiro4broad replied:
Not sure where this info is in the java spec, but `void` must be allowed as `Consumer` is a functional interface.

pshapiro4broad replied:
FYI here's the relevant part of the JLS https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.8

ddietterich replied:
Got it. It is a little bit of repetition in the DbRetry class, but gets rid of the ugly `return null` in the main code. Seems like a good trade-off.

wchamber commented on src/main/java/bio/terra/stairway/DbRetry.java:
```diff
@@ -0,0 +1,69 @@
+  private static final String PSQL_CONNECTION_ISSUE_PREFIX = "08";
+  private static final String PSQL_RESOURCE_ISSUE_PREFIX = "53";
+
+  private final String logString;
```
add a quick comment on what this is or should be used for

wchamber commented on src/main/java/bio/terra/stairway/DbUtils.java:
```diff
@@ -0,0 +1,23 @@
+import java.sql.SQLException;
+
+/** Database methods shared across DAOs */
+public class DbUtils {
+  static void startTransaction(Connection connection) throws SQLException {
```
Does this class need to be public if it's functions are package private? ditto on DbRetry

wchamber commented on src/main/java/bio/terra/stairway/StairwayInstanceDao.java:
```diff
@@ -0,0 +1,191 @@
+    justification = "Spotbugs doesn't understand resource try construct")
+public class StairwayInstanceDao {
+  private static final Logger logger = LoggerFactory.getLogger(StairwayInstanceDao.class);
+  static final String STAIRWAY_INSTANCE_TABLE = "stairwayinstance";
```
can this be private?

wchamber commented on src/main/java/bio/terra/stairway/StairwayInstanceDao.java:
```diff
@@ -0,0 +1,191 @@
+@SuppressFBWarnings(
+    value = "RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE",
+    justification = "Spotbugs doesn't understand resource try construct")
+public class StairwayInstanceDao {
```
ditto does this class need to be public or just package private?

ddietterich replied:
I made a pass through all of the classes to clean up public classes that shouldn't be. Thanks for picking up on that.

yonghaoy commented on src/main/java/bio/terra/stairway/DbRetry.java:
```diff
@@ -0,0 +1,69 @@
+  private static final int MAX_RETRIES = 20;
+  private static final int WAIT_MIN_MS = 250;
+  private static final int WAIT_MAX_MS = 1000;
+  private static final String PSQL_SERIALIZATION_FAILURE = "40001";
```
One of the exception I saw from log is 
`org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.` [[log]](org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.)
I didn't find the number for that exception. Does it covered by any of those code?

ddietterich replied:
I believe that will be a SQLstate beginning with PSQL_CONNECTION_ISSUE_PREFIX ("08").
The SQLstate is a member of the PSQLException object, so if you have logging that prints the whole object, you should have that value.

ddietterich committed:
General cleanup making public classes and methods package private
Use a static method for the retry so callers are simpler
Use a second method for retrying void methods so they don't need
to return Void/null
Address other review comments

ddietterich committed:
Restore the DuplicateFlightIdSubmittedException at the Stairway interface.
However, use the runtime DuplicateFlightIdException for throwing
within the code to avoid longer and longer throws lines.
